Css3 styling

jaytaph opened the pull request:
This (draft) PR implements a CSS system that allows to easily fetch properties from a renderer perspective.

We already are able to parse (internal and external) stylesheets into CSS ASTs. These are automatically attached to a document(handle).

From this point, a useragent can generate a render-tree which is a document but without any invisible or unrenderable nodes. This render tree should have for each node its CSS properties so you can easily find any CSS property on that node.

If there are no properties for a specific node, it will check if the queried property is inheritable (like font). if so, it will traverse down the nodes to find the property there. If none are found, or the property is not inheritable, it will use the default value for that property as defined.

In order to make sure that properties have correct values, we can use the [CSS documention's Value Syntax](https://developer.mozilla.org/en-US/docs/Web/CSS/Value_definition_syntax), where many properties are defined. This means that whenever we have a syntax for a property, we can check if any given value that is defined for that property actually matches the syntax (for instance: the color does not understand 3em).

The system is also able to generate shorthand properties. For instance, "border" is a property that is shorthand for border-color, border-size and border-style. These properties are again even shorthand for border-top-*, border-right-*, etc.

So when have defined a border: 1px solid black, we can fetch border-bottom-style, and retrieve the value "solid".

To do or to check:

- [X] Syntax compilation
- [ ] Value to syntax matching
- [ ] Inheritance
- [ ] Initial values
- [ ] Shorthand
- [ ] API to fetch values

Future work:

In a later PR, we will change this API to actually return enums that allows to do better matching. For instance, border-bottom-style property will return not a CssValue(), but a CssBorderStyle enum which could be something like:

```
enum CssBorderStyle {
   Solid,
   Striped,
   ...
   Width(float),
}
```

This way, we have the rust compiler to check if we are forgetting and not handling certain CSS3 values for given properties.

jaytaph committed:
Trying a new css system

jaytaph committed:
Rebased on main so we are updated again

jaytaph committed:
up

jaytaph committed:
adding scalars

jaytaph committed:
up

Sharktheone committed:
fix a bunch of test cases

Sharktheone committed:
implement list with option two

Sharktheone committed:
syntax matcher scalars, functions anr property

Sharktheone committed:
make the tests pass with a hack

Sharktheone committed:
fmt

Sharktheone committed:
fix lints

jaytaph committed:
added zero as separate cssvalue

jaytaph committed:
more fixes in the syntax matcher

Sharktheone reviewed (dismissed):
I don't know what I'm doing here... I wanted to see your latest changes, but my phone did things...

jaytaph committed:
simplified the group matching

jaytaph committed:
additional attempts to get multipliers correct

jaytaph committed:
juxtaposition seems to work, also with optional values

jaytaph committed:
Added some additional multiplier tests

Sharktheone committed:
make more of the test pass

jaytaph committed:
ran cargo fmt

jaytaph committed:
Fixed path for html5lib

jaytaph committed:
Updated rust version

jaytaph committed:
merged syntaxcomponent and syntaxcomponenttype

jaytaph committed:
Reverted back the FIXTURE_ROOT

jaytaph committed:
Removed debugging lines, and almost got a working test_prop

jaytaph committed:
up

Sharktheone committed:
fix some tests

Sharktheone committed:
fix the rest of the tests

jaytaph committed:
Add mdn syntax parser tool

jaytaph committed:
Initial updates to the css definition files

jaytaph committed:
Resolving syntax, almost all tests are passing again

jaytaph committed:
precedence is not correct, snapshot for fixing

jaytaph committed:
reversed the precedence, and trying to get juxtaposition working

Sharktheone committed:
initial go script for things... but there are still duplicated entries

Sharktheone committed:
move other files than main in main package to their own packages

Sharktheone committed:
add spec patcher

Sharktheone committed:
blacklist specs that cause problems for now

Sharktheone committed:
implement spec patching

Sharktheone committed:
download patches

jaytaph committed:
Precedence and groups seems to be working correct

Issues solved with juxtaposition which didn't match correctly.
Also, fixed precedence which was in the wrong order. THe only
thing still needs to do is fix the tests so they not only compile
the syntax, but also check the results (it didn't, so we didn't
catch the errors). It's a bit hard to do right now at my vacation
address with slow wifi, i'm longing for the days of a 2400 baud modem.

Sharktheone committed:
file patching finally working!

Sharktheone committed:
correctly handle files deleted by patches now

jaytaph committed:
Finished the big test

jaytaph committed:
Fixed a bunch more tests (precedences)

jaytaph committed:
argument lists are parsed correctly

Sharktheone committed:
write the properties to json files!

jaytaph committed:
added some debug stuff to test the matcher

jaytaph committed:
broke the matching issue down to the bare minimum test

Sharktheone committed:
it is working now!

Sharktheone committed:
fix units in datatype ranges

Sharktheone committed:
add our own definition files

Sharktheone committed:
remove old mdn tool

Sharktheone committed:
don't omit empty syntaxes for value

Sharktheone committed:
actually use our own files now

jaytaph committed:
new jutaposition setup

jaytaph committed:
multiple multipliers are allowed

jaytaph committed:
removed obsolete files

Sharktheone committed:
fix check_multiplier

Sharktheone committed:
remove `"`, `<`, `>`, `"` from value names

jaytaph committed:
optional multiplier works

jaytaph committed:
all multipliers seem to work well in juxtaposition

jaytaph committed:
working on exactlyone

jaytaph committed:
exactly one works.

jaytaph committed:
The four different group combinators seem to work

jaytaph committed:
Some updates on the definition generator

jaytaph committed:
got the definition syntax working again

jaytaph committed:
Update on the matcher

Sharktheone committed:
remove skip list, since it is no longer needed

jaytaph committed:
small fix where input was consumed when it shouldn't

jaytaph committed:
up

Sharktheone committed:
update definition files => removed skip list

Sharktheone committed:
fix match_exactly_one

Sharktheone committed:
THIS NEEDS TO BE REVERTED; Temporary fix: add `attr()` and `element()` to `BUILTIN_DATA_TYPES`

Sharktheone committed:
handle values that define values that define values

Sharktheone committed:
update definitions from last tool update

jaytaph committed:
up

jaytaph committed:
fixed the match_exactly_once grouping

jaytaph committed:
Refactored the multiplier tests into match_components

jaytaph committed:
All tests are passing now

jaytaph committed:
tests passing.. time for cleanup

jaytaph committed:
Cleaned up the css code

jaytaph committed:
some linting fixes

jaytaph committed:
test fix

Sharktheone commented on crates/gosub_styling/Cargo.toml:
```diff
@@ -15,9 +15,19 @@ lazy_static = "1.5"
 anyhow = "1.0.86"
 regex = "1.10.5"
 colors-transform = "0.2.11"
+vello = "0.1.0"
```
I think vello can be removed here. But I guess that's also happening when we resolve the current conflicts, so just a side note to check if it really is gone afterwards

Sharktheone commented on crates/gosub_styling/src/css_definitions.rs:
```diff
@@ -0,0 +1,851 @@
+    "x",
+    "y",
+    "color()",
+    "attr()",    //TODO: this is not a builtin!
+    "element()", //TODO: this is not a builtin!
```
Oh... true, we still need to see what we can do with these two guys...

Sharktheone commented on crates/gosub_styling/src/syntax.rs:
```diff
@@ -0,0 +1,2445 @@
+
+/// Represents an optional min and/or max range for a type definition
+#[derive(Clone, Debug, PartialEq)]
+pub struct RangeType {
+    /// Mininum value
+    min: NumberOrInfinity,
+    /// Maximum value
+    max: NumberOrInfinity,
+}
```
Maybe we can use rust's internal range type instead of our own?
